src/imperative/imperative_utils.cc

```diff
@@ -20,62 +20,61 @@
 #include "./imperative_utils.h"
 #include "./cached_op.h"
 
-namespace mxnet {
-namespace imperative {
+namespace {
 
-inline std::vector<NDArray*> NodeInputs(const nnvm::IndexedGraph& idx,
-                                        const int node_idx,
-                                        const std::vector<NDArray*> arrays) {
+std::vector<NDArray*> NodeInputs(const nnvm::IndexedGraph& idx,
```

Comment:
Can we pass in the returned vector `std::vector<NDArray*>` so we can save another copy?

Reply:
This would be unnecessary and non idiomatic. Return value optimization (RVO) will kick in and eliminate any extra copies on return values. 

https://shaharmike.com/cpp/rvo/